src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -16550,11 +16551,22 @@ def str_len(self, **kwargs: Any) -> "SnowflakeQueryCompiler":
         SnowflakeQueryCompiler representing result of the string operation.
         """
         # TODO SNOW-1438001: Handle dict, list, and tuple values for Series.str.len().
-        return SnowflakeQueryCompiler(
-            self._modin_frame.apply_snowpark_function_to_columns(
-                lambda col: self._replace_non_str(col, length(col))
+        col = self._modin_frame.data_column_snowflake_quoted_identifiers[0]
+        if isinstance(
+            self._modin_frame.quoted_identifier_to_snowflake_type([col]).get(col),
+            ArrayType,
+        ):
+            return SnowflakeQueryCompiler(
+                self._modin_frame.apply_snowpark_function_to_columns(
+                    lambda col: array_size(col)
+                )
+            )
+        else:
```

Comment:
since you are doing the type check anyway, how about just do a check of string type and throw an unsupported error for other cases ? length is only working for string, right?

Reply:
I added the NotImplementedError but then had to remove it again because I realized that we do support a special case of mixed types (strings + numerical values, where the numerical values are automatically coerced into strings by snowflake). These test cases would break if we only allow string or array column types.